src/sentry/auth/helper.py

```diff
@@ -551,104 +504,64 @@ class AuthHelper:
     # configuring the provider
     FLOW_SETUP_PROVIDER = 2
 
+    pipeline_name = "pipeline"
+    provider_manager = manager
+    provider_model_cls = AuthProvider
+    session_store_cls = RedisBackedState
+
     @classmethod
     def get_for_request(cls, request):
-        state = RedisBackedState(request)
-        if not state.is_valid():
+        req_state = cls.unpack_state(request)
+        if not req_state:
             return None
 
-        organization_id = state.org_id
-        if not organization_id:
+        if not req_state.organization:
             logging.info("Invalid SSO data found")
             return None
-
-        flow = state.flow
-        auth_provider_id = state.auth_provider
-        provider_key = state.provider_key
-        if auth_provider_id:
-            auth_provider = AuthProvider.objects.get(id=auth_provider_id)
-        elif provider_key:
-            auth_provider = None
-
-        organization = Organization.objects.get(id=state.org_id)
+        flow = req_state.state.flow
 
         return cls(
-            request, organization, flow, auth_provider=auth_provider, provider_key=provider_key
+            request,
+            req_state.organization,
+            flow,
+            auth_provider=req_state.provider_model,
+            provider_key=req_state.provider_key,
```

Comment:
`get_for_request` is the one of a few `Pipeline` methods that `AuthHelper` overrides entirely, rather than dispatching to `super`. This is a concession to having a completely different `__init__` signature; neither `__init__` method is changed.

I extracted the new `unpack_state` method to avoid as much code duplication as possible.